OWML.ModHelper.Input/ModInputCombination.cs

```diff
@@ -0,0 +1,159 @@
+        {
+            switch (xboxKey[0])
+            {
+                case 'A':
+                    return InputTranslator.GetButtonKeyCode(JoystickButton.FaceDown);
+                case 'B':
+                    return InputTranslator.GetButtonKeyCode(JoystickButton.FaceRight);
+                case 'X':
+                    return InputTranslator.GetButtonKeyCode(JoystickButton.FaceLeft);
+                case 'Y':
+                    return InputTranslator.GetButtonKeyCode(JoystickButton.FaceUp);
+                default:
+                    return StringToKeyCodeGamepad(xboxKey);
+            }
+        }
+
+        private KeyCode StringToKeyCode(string key)
+        {
+            key = key.Trim();
+            return key.Contains(XboxPrefix) ? StringToKeyCodeXbox(key.Substring(XboxPrefix.Length)) 
+                : StringToKeyCodeKeyboard(key);
+        }
+
+        private int[] StringToKeyArray(string stringCombination)
+        {
+            var keyCombination = new int[MaxComboLength];
+            var i = 0;
+            foreach (var key in stringCombination.Trim().ToLower().Split('+'))
+            {
+                var code = StringToKeyCode(key);
+                if (code == KeyCode.None)
+                {
+                    keyCombination[0] = (int)RegistrationCode.InvalidCombination;
+                    return keyCombination;
+                }
+                if (i >= MaxComboLength)
+                {
+                    keyCombination[0] = (int)RegistrationCode.CombinationTooLong;
+                    return keyCombination;
+                }
+                keyCombination[i] = (int)code;
+                i++;
+            }
+            Array.Sort(keyCombination);
+            return keyCombination;
+        }
+
+        private long StringToHash(string stringCombination)
+        {
+            var keyCombination = StringToKeyArray(stringCombination);
+            if (keyCombination[0] < 0)
+            {
+                return keyCombination[0];
+            }
+            long hash = 0;
+            for (var i = 0; i < MaxComboLength; i++)
+            {
+                hash = hash * MaxUsefulKey + keyCombination[i];
+            }
+            return hash;
+        }
+
+        private List<long> StringToHashes(string combinations)
+        {
+            var hashes = new List<long>();
+            foreach (var combo in combinations.Split('/'))
+            {
+                var hash = StringToHash(combo);
+                if (hash <= 0)
+                {
+                    hashes.Clear();
+                    hashes.Add(hash);
+                    return hashes;
+                }
+                hashes.Add(hash);
+                if (hash < MaxUsefulKey)
+                {
+                    _singles.Add((KeyCode)hash);
+                }    
+            }
+            return hashes;
+        }
+
+        internal ModInputCombination(IModManifest mod, string name, string combination)
```

Comment:
ah, here's the constructor. To the top :)

Reply:
done